### PR TITLE
Add note of CVE-2023-4039 to -fstack-protector description

### LINKED ISSUE #234: GCC's -fstack-protector fails to guard dynamic stack allocations on Aarch64
On September 12th, 2023 [Meta's Read Team disclosed a vulnerability in GCC on AArch64 targets](https://github.com/metaredteam/external-disclosures/security/advisories/GHSA-x7ch-h5rf-w2mf) tracked as [CVE-2023-40339](https://nvd.nist.gov/vuln/detail/CVE-2023-40339) that causes GCC's stack smashing protection not to detect or defend against overflows of dynamically-sized local variables such as variable-length arrays or buffers allocated using `alloca()`.

At the time of disclosure, all version of GCC from 5.4.0 to trunk as of 2023-05-15 were affected. 

In response, [Arm issued an advisory](https://developer.arm.com/Arm%20Security%20Center/GCC%20Stack%20Protector%20Vulnerability%20AArch64) containing a patch that has been backported to GCC 7 - 13 and are incorporated to new releases of Arm GNU Toolchain.

Given that the Compiler Options Hardening Guide for C and C++ Guide recommends the use of GCC's stack smashing protection the guide should address this to ensure readers on Aarch64 targets benefit from stack smashing protection using patched version of GCC.

### PR DESCRIPTION
**Pull request by thomasnyman**
Fixes #234

### CODE REVIEW

**Comment by ware** [comment]
Is there a reason we're putting this in other than telling people to just use the latest, supported release?

**Comment by siddhesh** [comment]
Is this within the scope of the guide though?  In terms of best practices, one must always use at least `-fstack-protector-strong` and what the compiler does under the hood shouldn't really change that.

This creeps into the area of providing advisories for compilers (and the idea of providing advisories for missed hardening is dubious anyway, we'd be filing them till the cows mutate and go back into the ocean), which doesn't seem to fit in with this guide IMO.

**Comment by edelsohn** [comment]
Why is there a pull request for this change without addressing the objections from multiple people in the issue?

**Comment by thomasnyman** [comment]
Addressed the feedback to #234 based on the discussion at the [C/C++ Compiler BP Call 2023-09.27](https://docs.google.com/document/d/1UClGUnOSkOH_wab6Lx43KUdkaK37L8sbWJ_GPZvc1YY/edit#bookmark=id.863n4hp4nw9q).

The reworded change addressed the underlying issue as a failure in the `-fstack-protector[-strong,-all,]` feature functioning perfectly in all situations rather than as an immediate security issue, addresses what users of the compiler can do to obtain better results, and adds a general disclaimer earlier in the guide on compiler features that are  dependent on heuristics and may not always provide full coverage of protection.

More feedback on the reworded text is welcome.

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md [hunk 1]
```diff
@@ -376,6 +380,10 @@ The performance overhead is dependent on the number of function’s instrumented
 
 [^Han11]: Shen, Han, [New stack protector option for gcc](https://docs.google.com/document/d/1xXBH6rRZue4f296vGt9YQcuLVQHeE516stHwt8M9xyU), Google Docs, 2011-11-30.
 
+[^Meta23]: Hebb, Tom, [GCC's -fstack-protector fails to guard dynamic stack allocations on ARM64](https://github.com/metaredteam/external-disclosures/security/advisories/GHSA-x7ch-h5rf-w2mf), GithHu metaredteam/external-disclosures Advisories, 2023-09-12.
```
Comment (siddhesh) [inline docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:383]: s/GithHu/GitHub/

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md [hunk 2]
```diff
@@ -53,7 +53,9 @@ Consequently, modern operating systems deploy various run-time mechanisms to pro
...
 
-Some mechanisms may require additional configuration and fine tuning, for example due to potential compilation issues for certain unlikely edge cases, or performance overhead the mitigation adds for certain program constructs. This problem is exacerbated in projects that rely on an outdated version of an open source software (OSS) compiler. In general, security mitigations are more likely to be enabled by default in modern versions of compilers included with Linux distributions. Note that the defaults used by the upstream GCC project do not enable some of these mitigations.
+Some mechanisms may require additional configuration and fine tuning, for example due to potential compilation issues for certain unlikely edge cases, or performance overhead the mitigation adds for certain program constructs. Some compiler features are dependent on heuristics that determine when the protection mechanism is applied and may not always provide full coverage of protection.
```
Comment (siddhesh) [inline docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:56]: How about this instead:

> Compiler security features depend on data flow analysis of programs and heuristics, results of which may vary depending on program source code layout. As a result, the protection mechanisms implemented by these features may not always provide full coverage.

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md [hunk 3]
```diff
@@ -355,6 +357,8 @@ The detection is based on inserting a *canary* value into the stack frame in the
 
 This mitigates potential control-flow hijacking attacks that may lead to arbitrary code execution by corrupting return addresses stored on the stack.
 
+Out-of-date versions of GCC may not detect or defend against overflows of dynamically-sized local variables such as variable-length arrays or buffers allocated using `alloca()` when compiling for 64-bit Arm (Aarch64) processors[^Meta23]. Users of GCC-based toolchains for Aarch64 should ensure they use up-to-date versions of GCC 7 or later that support an alternative stack frame layout that places all local variables below saved registers, with the stack-protector canary between them[^Arm23].
```
Comment (siddhesh) [inline docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:360]: I just noticed that there's no release with the aarch64 patchset; the patchset was merely pushed into release branches all the way back to gcc-7. Should we wait until there's an actual release before this note is added?  For example, I don't know if a gcc-7 update release will even happen at all; the last one happened in 2019.

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md (hunk 2 above)
Comment (thomasnyman) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:56]: layout -> details

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md (hunk 3 above)
Comment (david-a-wheeler) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:360]: Possible reword:


> Some versions of GCC may not detect or defend against overflows of dynamically-sized local variables such as variable-length arrays or buffers allocated using `alloca()` when compiling for 64-bit Arm (Aarch64) processors[^Meta23]. Users of GCC-based toolchains for Aarch64 should, before depending on it, determine if they support an alternative stack frame layout that places all local variables below saved registers, with the stack-protector canary between them[^Arm23].

Comment (thomasnyman) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:360]: I've incorporated the suggested wording in 2763a9c3e5079000859017437d01806754f6a56b.

**Comment by thomasnyman** [comment]
Addressed feedback from [C/C++ BP Call on 2023-10-11](https://docs.google.com/document/d/1UClGUnOSkOH_wab6Lx43KUdkaK37L8sbWJ_GPZvc1YY/edit#bookmark=id.97ccyuu7twe).

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md (hunk 1 above)
Comment (david-a-wheeler) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md]: Caps fixed!

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md (hunk 2 above)
Comment (david-a-wheeler) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md]: I think this is resolved. Let me know if that's not correct.

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md (hunk 3 above)
Comment (david-a-wheeler) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:360]: Sounds like this is separately resolved.

**Comment by thomasnyman** [comment]
Merging based on discussion in [C/C++ BP Call on 2023-10-11] (https://docs.google.com/document/d/1UClGUnOSkOH_wab6Lx43KUdkaK37L8sbWJ_GPZvc1YY/edit#bookmark=id.97ccyuu7twe) as no objections on the new wording has been raised.

**Pull request merged**